### PR TITLE
Add Brotli4J compression related classes, files and dependencies

### PR DESCRIPTION
**Pull request by jcarranzan**
### Summary

    Integration of Brotli4J compression tests with Quarkus REST endpoints.
    Test Plan related to: [QQE-378](https://github.com/quarkus-qe/quarkus-test-plans/blob/main/QQE-378.md)

Test Coverage:

    JVM and native mode testing will be included.

Test Scenarios Covered:

-     Compression Enabled/Disabled:

        Confirms response compression with Brotli when enabled/disabled.

-     Media Types size Payloads:

        Verifies Brotli compression up to 8192 bytes for JSON payloads.
        Validates error handling for payloads larger than 8192 bytes.

-     Different Media Types of payload compression (text, JSON, XML):

        Checks compression correctness 

Please select the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Dependency update
- [ ] Refactoring
- [ ] Backport
- [x] New scenario (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
- [ ] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [x] Methods and classes used in PR scenarios are meaningful
- [x] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

### CODE REVIEW

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+               .extract().response();
+        byte[] responseBody = response.getBody().asByteArray();
+        int compressedContentLength = responseBody.length;
+        assertTrue(compressedContentLength < originalContentLength);
+
+    }
+
+    @Test
+    public void disableCompression(){
+        System.setProperty("quarkus.http.enable-compression", "false");
```
Comment (jedla97) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java:79]: I'm not sure if this is correct usage with quarkus. Maybe it is but can't say what should be correct way.

On another note, how we ensuring that this property is only set for this test and not affecting others test?

Also I tried this test, removed this property and the test still pass. Shouldn't this fail when compression is disabled?

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    private int calculateOriginalJsonLength(String jsonFilePath) throws IOException {
+        try (InputStream inputStream = getClass().getResourceAsStream(jsonFilePath)) {
+            ObjectMapper objectMapper = new ObjectMapper();
+            JsonNode jsonNode = objectMapper.readTree(inputStream);
+            String jsonString = objectMapper.writeValueAsString(jsonNode);
+            return jsonString.getBytes().length;
+        }
+    }
+
+    private int calculateXmlLength() throws IOException {
```
Comment (jedla97) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java:149]: Wouldn't be better have the xml file path as parameter same as in `calculateOriginalJsonLength` method?

**Review by jedla97 — CHANGES_REQUESTED** [review]
Thanks for adding coverage @jcarranzan. I have some comments most of them are minor.

Looking at the test plan I see that you plan to test it with openshift. Will this testcoverage part of this PR or it will be separate PR? I think it can be part of this PR but leave the decision up to you

Also can you modify commit message not to include squashed commit messages to keep it more clean.

**Comment by jcarranzan** [comment]
> Thanks for adding coverage @jcarranzan. I have some comments most of them are minor.
> 
> Looking at the test plan I see that you plan to test it with openshift. Will this testcoverage part of this PR or it will be separate PR? I think it can be part of this PR but leave the decision up to you
> 
> Also can you modify commit message not to include squashed commit messages to keep it more clean.

Hi, yes I will include Openshift coverage in this PR, and also another test about big file compression (>8192 bytes), I 'll let you know it when is ready.
Also, I'll clean squashed commit message, thanks @jedla97 .

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JRestMock.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            return xmlResponse;
+        } else {
+            throw new IllegalArgumentException("Invalid response type: " + type);
+        }
+    }
+  public enum ResponseType {
+        JSON, XML
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JRestMock.java:129]: why?

[the remark this answers is not in the document]
Comment (jedla97) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JRestMock.java]: It's more like a standard to end files with new line. As what I remember some tools can expect this + I thinks it's part of some definition  (not sure)

[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JRestMock.java]: Ok thanks. I will try to configure my IDE to add one line at the end. Or I will apply (if exist) some code style guides file to do it.

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        try (InputStream inputStream = getClass().getResourceAsStream("/xml_sample.xml")) {
+            Scanner scanner = new Scanner(inputStream, StandardCharsets.UTF_8);
+            StringBuilder stringBuilder = new StringBuilder();
+            while (scanner.hasNextLine()) {
+                stringBuilder.append(scanner.nextLine()).append("\n");
+            }
+            return stringBuilder.toString().length();
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java:173]: Yes, it is :)

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                .body(is(DEFAULT_TEXT_PLAIN)).log().all();
+    }
+
+    @Test
+    public void checkTextPlainWithtBrotli4J() throws IOException {
+        assertBrotli4JCompression("/compression/text", MediaType.TEXT_PLAIN, BROTLI_ENCODING, BROTLI_ENCODING, CONTENT_LENGTH_DEFAULT_TEXT_PLAIN);
+
+    }
+
+    public void assertBrotli4JCompression(String path, String contentHeaderType, String acceptHeaderEncoding, String expectedHeaderContentEncoding, int originalContentLength) {
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java]: Yes, I will move there, thanks.

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpServerConfig.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+import io.netty.handler.codec.compression.BrotliOptions;
+import io.netty.handler.codec.compression.StandardCompressionOptions;
+import io.quarkus.runtime.Startup;
+import io.quarkus.vertx.http.HttpServerOptionsCustomizer;
+import io.vertx.core.http.HttpServerOptions;
+
+@Startup
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpServerConfig.java:11]: When I started to develop this code, I needed that the Brotli4JHttpServer configuration with all the customized brotli options were set at the beginning of the lifecycle application, and I checked that way was doing this: https://quarkus.io/guides/lifecycle#startup_annotation  .
After your comment, I tested without the annotation, and works fine also, so I guess Quarkus does the work with no need of @Startup annotation, so I will remove it. Thanks.

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpServerConfig.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+@ApplicationScoped
+public class Brotli4JHttpServerConfig implements HttpServerOptionsCustomizer {
+    // It depends on compression level that we want apply
+    private final int compressionLevel = 4;
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpServerConfig.java:14]: I tested manually, and it depends on the level of compression whether one or the other is used, resulting in more or fewer compressed bytes in length, as well as the speed ratio level, which depends on the assigned level. Look at the differences, for instance, in the Compression Levels section here:  https://paulcalvano.com/2018-07-25-brotli-compression-how-much-will-it-reduce-your-content/

[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpServerConfig.java:14]: I will put static variable yes.

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java (hunk 1 above)
Comment (jcarranzan) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java:71]: Yeah, good point @jedla97 and thank you for the explanation @michalvavrik .
I will fix it, sorry about that guys.

Comment (jcarranzan) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java]: done with @DevModeQuarkusApplication

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+                .contentType(contentHeaderType)
+                .header(HttpHeaders.ACCEPT_ENCODING, acceptHeaderEncoding)
+                .get(path)
+                .then()
+                .statusCode(200)
+                .header(HttpHeaders.CONTENT_ENCODING, expectedHeaderContentEncoding)
+                .extract().response();
+        byte[] responseBody = response.getBody().asByteArray();
+        int compressedContentLength = responseBody.length;
+        assertTrue(compressedContentLength < originalContentLength);
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java]: Currently, with the disableCompression test (in Dev Mode) I verify the content length header is the same for the default text, IMHO it's enough.

**Comment by jcarranzan** [comment]
> * You have marked `This change requires a documentation update` but there is no docs update. Please add docs update.
> 
>     * You have marked `Dependency update`, I'll respect your choice, but I think it's meant for a bump. I am not sure I'd call adding dep an update.

Unmarked checks are not needed for this coverage. Thanks

**Comment by jcarranzan** [comment]
run tests

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/DevModeBrotli4JHttpIT.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+import io.quarkus.test.bootstrap.DevModeQuarkusService;
+import io.quarkus.test.scenarios.QuarkusScenario;
+import io.quarkus.test.services.DevModeQuarkusApplication;
+import io.restassured.response.Response;
+
+@QuarkusScenario
+public class DevModeBrotli4JHttpIT {
+    private static final String COMPRESSION_ENABLED_PROPERTY = "quarkus.http.enable-compression";
+    @DevModeQuarkusApplication(classes = { Brotli4JHttpServerConfig.class, Brotli4JResource.class,
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/DevModeBrotli4JHttpIT.java:23]: Well, in this case for just one test to change the app setup, from my point of view, the approach with DEV mode is correct (also this idea was commented on here: https://github.com/quarkus-qe/quarkus-test-suite/pull/1717#discussion_r1588403454 ), if we need more extra tests with compression disabled then it is worth using another properties file configuration for a new one specific test class.

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            String jsonString = objectMapper.writeValueAsString(jsonNode);
+            return jsonString.getBytes().length;
+        }
+    }
+
+    private int calculateXmlLength(String xmlFilePath) {
+        return FileUtils.loadFile("/sample.xml").length();
+    }
+
+    private int calculateTextLength(String fileName) {
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java:148]: It's the same, there are different ways to do the same, I did it as a standard way to read the textile...
But I can do it as in calculateXmlLength to keep some kind of coherence in all the methods. Thanks.

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        try (InputStream inputStream = getClass().getResourceAsStream(jsonFilePath)) {
+            ObjectMapper objectMapper = new ObjectMapper();
+            JsonNode jsonNode = objectMapper.readTree(inputStream);
+            String jsonString = objectMapper.writeValueAsString(jsonNode);
+            return jsonString.getBytes().length;
+        }
+    }
+
+    private int calculateXmlLength(String xmlFilePath) {
+        return FileUtils.loadFile("/sample.xml").length();
```
[the remark this answers is not in the document]
Comment (jcarranzan) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Brotli4JHttpIT.java:145]: I will put it inline inside the method

**Comment by jcarranzan** [comment]
run tests

**Comment by jcarranzan** [comment]
run tests

**Comment by jedla97** [comment]
@jcarranzan Looking at this it's loog good. Just one question as the https://github.com/quarkusio/quarkus/issues/40533 is not just on openshift but also occuring on baremetal shouldn't be `Brotli4JHttpIT` also disabled on native?

I run it localy using `mvn clean verify -Phttp-modules -Dfailsafe.failIfNoSpecifiedTests=false -pl http/http-advanced-reactive -Dreruns=0 -Dit.test="Brotli4JHttpIT" -Dnative` and hitting the error. Looking at the our CI for native and it didn't execute any tests (don't know why) https://github.com/quarkus-qe/quarkus-test-suite/actions/runs/9016177799/job/24773143474?pr=1717

**Comment by jcarranzan** [comment]
> @jcarranzan Looking at this it's loog good. Just one question as the [quarkusio/quarkus#40533](https://github.com/quarkusio/quarkus/issues/40533) is not just on openshift but also occuring on baremetal shouldn't be `Brotli4JHttpIT` also disabled on native?
> 
> I run it localy using `mvn clean verify -Phttp-modules -Dfailsafe.failIfNoSpecifiedTests=false -pl http/http-advanced-reactive -Dreruns=0 -Dit.test="Brotli4JHttpIT" -Dnative` and hitting the error. Looking at the our CI for native and it didn't execute any tests (don't know why) https://github.com/quarkus-qe/quarkus-test-suite/actions/runs/9016177799/job/24773143474?pr=1717

Yes, it should be disabled also on baremetal in this case. I will push my changes right now.
On the other hand @jedla97 , IDK either why any tests were executed, maybe any external change in our ts could have affected this?

**Review by jedla97 — APPROVED** [review]
Thanks for update!. Now it look good.

**Comment by jedla97** [comment]
CI is green + native modules was executed this time so merging

**Pull request merged**